# Exercise 2.A Fundamentals of pandas

In [25]:
import numpy as np

In [26]:
import pandas as pd

## Series 1

In [27]:
aa = [1, 2, 3, 4]

In [28]:
ser1 = pd.Series(aa)

In [29]:
print(ser1)

0    1
1    2
2    3
3    4
dtype: int64


## Series 2

In [30]:
ab = (5, 6, 7, 8)

In [31]:
ser2 = pd.Series(ab)

In [32]:
print(ser2)

0    5
1    6
2    7
3    8
dtype: int64


## Series 3

In [33]:
ac = list({9, 10, 11, 12})

In [34]:
ser3 = pd.Series(ac)

In [35]:
print(ser3)

0     9
1    10
2    11
3    12
dtype: int64


## Series 4

In [36]:
ad = {'first': 13, 'second': 14, 'third': 15, 'fourth': 16} 

In [37]:
ser4 = pd.Series(ad)

In [38]:
print(ser4)

first     13
second    14
third     15
fourth    16
dtype: int64


## Series 5

ser5 = pd.Series(17, 18, 19, 20)

Type error is "index(...)

## DataFrames from Series

In [40]:
df1 = pd.DataFrame(ser1)

In [42]:
df1

,0
0,1
1,2
2,3
3,4


In [46]:
df1.insert(1, "Column2", ser2)

In [48]:
df1.insert(2, "Column3", ser3)

In [49]:
df1

,0,Column2,Column3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [50]:
df1.insert(3, "Column4", ser4)

In [53]:
df1

,0,Column2,Column3,Column4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [54]:
# ser4 uses dictionary keys (first, second, third, fourth)
# as its index instead of numeric indexes (0,1,2,3).
# Because the indexes do not match the DataFrame's indexes,
# pandas aligns by index, producing NaN values.

In [55]:
df1.loc[4] = [5, 9, 12, 0]

In [56]:
df1

,0,Column2,Column3,Column4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [57]:
df2 = pd.read_csv('sakila_example.csv')

In [60]:
df2.head(5)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [61]:
df2.head(10)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
5,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,MENA TEMPLE
6,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,WARREN NOLTE
7,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,OPRAH KILMER
8,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,ROCK DUKAKIS
9,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,MARY KEITEL


In [64]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


### Insights from info()

The dataset contains 262 records and 5 text-based columns: title, description, category, language, and actor_name. Since all columns are strings, the most useful information from describe() is the count, unique, top, and frequency values. These statistics help identify whether there are duplicate values, the most common categories, languages, or actors, and how diverse the data is.

Some summarizations are less useful. The title and description fields are likely intended to be unique for most records, so knowing the most common title or description does not provide much analytical value. Since all columns are text-based, there are no numerical statistics such as mean, median, minimum, or maximum available.

The file appears to be unnormalized because actor names, categories, and languages are stored directly within the same table as the movie information. In a normalized database design, these attributes would likely be separated into related tables and connected through keys to reduce redundancy and improve data integrity.


In [68]:
df3 = pd.read_csv('northwind_example.csv')
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [69]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


This file is unnormalized but does give good information.
It looks like it is giving a detalied comparison and ranking of the information in the file

## Coding a DataFrame

In [70]:
test_data = {'ID' : [1, 2, 3, 4, 5 ,6 ,7 ,8], 
             'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'], 
             'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18], 
             'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'], 
             'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]}

In [72]:
df4 = pd.DataFrame(test_data)

In [73]:
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [74]:
df4.to_csv('test_data.csv', index=False)